packages/react-data/src/fetch/ReactDataFetch.js

```diff
@@ -44,50 +46,114 @@ function readResultMap(): Map<string, Result> {
-      const resolvedResult = ((result: any): ResolvedResult);
-      resolvedResult.status = Resolved;
-      resolvedResult.value = xhr.response;
-      resolve();
-    };
-    xhr.onerror = function() {
-      if (result.status !== Pending) {
-        return;
+    },
+    err => {
+      if (result.status === Pending) {
+        const rejectedResult = ((result: any): RejectedResult);
+        rejectedResult.status = Rejected;
+        rejectedResult.value = err;
       }
-      const rejectedResult = ((result: any): RejectedResult);
-      rejectedResult.status = Rejected;
-      // TODO: use something else as the error value?
-      rejectedResult.value = xhr;
-      resolve();
-    };
-    xhr.open('GET', url);
-    xhr.send();
-    throw wakeable;
-  }
-  const result: Result = entry;
+    },
+  );
+  return result;
+}
+
+function readResult(result: Result) {
   if (result.status === Resolved) {
     return result.value;
   } else {
     throw result.value;
   }
 }
+
+function getArrayBuffer(nativeResponse) {
+  return nativeResponse.arrayBuffer();
+}
+
+function getBlob(nativeResponse) {
+  return nativeResponse.blob();
+}
+
+function getJson(nativeResponse) {
+  return nativeResponse.json();
+}
+
+function getText(nativeResponse) {
+  return nativeResponse.json();
+}
+
+function Response(nativeResponse) {
+  this.headers = nativeResponse.headers;
+  this.ok = nativeResponse.ok;
+  this.redirected = nativeResponse.redirected;
+  this.status = nativeResponse.status;
+  this.statusText = nativeResponse.statusText;
+  this.type = nativeResponse.type;
+  this.url = nativeResponse.url;
+
+  this._entry = null;
+  this._consumed = null;
+  this._response = nativeResponse;
+}
+
+Response.prototype = {
+  constructor: Response,
+  arrayBuffer() {
+    return this._read('arrayBuffer', getArrayBuffer);
+  },
+  blob() {
+    return this._read('blob', getBlob);
+  },
+  json() {
+    return this._read('json', getJson);
+  },
+  text() {
+    return this._read('text', getText);
+  },
+  _read(type: string, getThenable) {
+    const consumed = this._consumed;
+    if (consumed != null && type !== consumed) {
+      throw new Error('Already read.');
+    }
+    let entry = this._entry;
+    if (entry == null) {
+      this._consumed = type;
+      const thenable = getThenable(this._response);
+      entry = this._entry = createFromThenable(thenable, r => r);
+    }
+    return readResult(entry);
+  },
+};
+
+export function fetch(url: string, options: mixed): Object {
+  const map = readResultMap();
+  let entry = map.get(url);
+  if (entry == null) {
+    if (options) {
+      if (options.method || options.body || options.signal) {
+        // TODO: wire up our own cancellation mechanism.
+        // TODO: figure out what to do with POST.
+        throw Error('Unsupported option');
+      }
+    }
+    const thenable = nativeFetch(url, options);
+    entry = createFromThenable(
+      thenable,
+      nativeResponse => new Response(nativeResponse),
```

Comment:
The way I think of these async points is that it gets simpler if everything is just at the deepest point. Then you can just use Suspense on top of that.

E.g. this is just like a `let nativeResponse = read(url); return new Response(nativeResponse)` like any other wrapper object around this would do in a suspensey function.

The difference is that you want to also memoize that object. Memoizing in this world is just passing deps into a resource and getting value out.

```
let nativeResponse =  readFetch(url);
return readResponse(nativeResponse);
```

I.e. there's a cache from nativeResponse to Response.

You could implement that as an expando on the nativeResponse.

```
let nativeResponse = readResult(entry);
if (nativeResponse._reactResponse) {
  return nativeResponse._reactResponse;
} else {
  return nativeResponse._reactResponse = new Response(nativeResponse);
}
```

That way you avoid the closure and it's a pattern you can use for many cases even if you need the underlying value for many different wrappers.